### ISSUE TITLE
Gracefully handle the absence of a versioning system

### ISSUE THREAD

**Issue by MarenstainBear** [brandon-rhodes/uncommitted#26]
Thoughts on handling the absence of a particular versioning systems?

Here's the issue I ran into:
```
/home/mary - Git
 M .config/terminator/config
 M .vimrc
* master b8212ae [ahead 1] More vim changes and also a window manager keymap change.

Traceback (most recent call last):
  File "/usr/local/bin/uncommitted", line 11, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.5/dist-packages/uncommitted/command.py", line 233, in main
    scan(repos, options)
  File "/usr/local/lib/python3.5/dist-packages/uncommitted/command.py", line 158, in scan
    lines, subrepos = get_status(directory, ignore_set, options)
  File "/usr/local/lib/python3.5/dist-packages/uncommitted/command.py", line 122, in status_subversion
    for line in run(['svn', 'st', '-v'], cwd=path):
  File "/usr/local/lib/python3.5/dist-packages/uncommitted/command.py", line 33, in run
    return check_output(command, **kw).splitlines()
  File "/usr/lib/python3.5/subprocess.py", line 626, in check_output
    **kwargs).stdout
  File "/usr/lib/python3.5/subprocess.py", line 693, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/usr/lib/python3.5/subprocess.py", line 947, in __init__
    restore_signals, start_new_session)
  File "/usr/lib/python3.5/subprocess.py", line 1551, in _execute_child
    raise child_exception_type(errno_num, err_msg)
FileNotFoundError: [Errno 2] No such file or directory: 'svn
```
And this seems to be where the problem originated for me:
[Lines 137:141 of command.py](https://github.com/brandon-rhodes/uncommitted/blob/90287c84598a4e1a1d5115561ded736ef43ddd66/uncommitted/command.py#L137)

I don't have SVN or hg, so it seems that *uncommitted* is failing when it expects an SVN directory, but doesn't find one. I worked around it by simply commenting out lines 139 & 140.

Some ideas to fix or enhance this are:

1. simply handle the absence of the SVN folder more gracefully. In other words, don't error out and instead continue processing any directories with recognized version control
2. allow the user the ability to select which versioning systems to scan for; OR,
3. conversely, allow the user to opt-out of specified versioning systems.

I'm happy to write a pull request to do all or some of the above. Thoughts?

**Comment by brandon-rhodes** [comment]
A quick clarification: it sounds, from your description, as though the problem is actually the presence of an `.svn` directory, but in the absence of the command line tool? That uncommitted has found a directory named `.svn` and wants to run the command but it isn't there?

A quick way to confirm would be to add a `print(directory, dotdir)` as the second line of the `while repos:` loop and see what directory name it prints out before it tries calling the `svn` function down in line 158.

At the very least, we'll want `uncommitted` to recover and keep going if a command fails to execute. If you expanded `except CalledProcessError` to also catch `FileNotFoundError`, does that allow `uncommitted` keep running?

**Comment by MarenstainBear** [comment]
Ah, yes, it was because of an existing .svn directory in my file system (see screenshot below). The program kept running when I caught (and ignored) the ```FileNotFoundError``` in the ```run(command, **kw)``` function like this:
```
def run(command, **kw):
    """Run `command`, catch any exception, and return lines of output."""
    try:
        # In Python 3, iterating over bytes yield integers, so we call
        # `splitlines()` to force Python 3 to give us lines instead.
        return check_output(command, **kw).splitlines()
    except CalledProcessError:
        return ()
    except FileNotFoundError:
        return ()
```

![image](https://user-images.githubusercontent.com/33665154/36316711-a752fa72-1300-11e8-988a-99813f298668.png)

**Comment by brandon-rhodes** [comment]
Nice! Would you like to offer a pull request to get credit for solving this bug you ran into? If so, just throw a print into your `except FileNotFoundError:` so that uncommitted will report to its users that it had to skip a checkout because a binary that it needed was missing — and, in the `print`, report which binary was missing (`svn` or whatever) and which directory was skipped (so the user can go look at it).

Or, I can look for some time to jump in and fix the code myself — but it would be fun if you contributed this fix since you're the one who discovered the failure mode for us!

**Comment by MarenstainBear** [comment]
Oh, I'd love to write a pull request for this! It'll be my first contribution to the open source community. :-D

I'll get a request off to you in the next day or so. Thanks!

**Comment by MarenstainBear** [comment]
Alright, pull request is out there. #27

**Comment by brandon-rhodes** [comment]
Thank you! I've made a small comment on it and, with that last tweak, your first contribution will be ready to land :)

**Comment by MarenstainBear** [comment]
Woot! Thanks for letting me collaborate with you. :-)

**Comment by brandon-rhodes** [comment]
Thanks for the contribution! :)

I just clicked on your profile to see the little green squares on the calendar showing this activity, and learned for the first time that GitHub has a special display for the moment when someone opens their first pull request. Congratulations. :)

![first-contribution](https://user-images.githubusercontent.com/166162/36461355-efa0cbfa-168a-11e8-8044-2c24d08a61a9.png)

**Comment by brandon-rhodes** [comment]
I've just released a new version 2.1 of uncommitted which should include this improvement. Enjoy! :)
